src/util-memcap.c

```diff
@@ -0,0 +1,184 @@
+        return 0;
+    }
+    new_memcap->name = name;
+    new_memcap->option = option;
+    new_memcap->SetFunc = SetFunc;
+    new_memcap->GetFunc = GetFunc;
+    new_memcap->GetMemuseFunc = GetMemuseFunc;
+    new_memcap->next = NULL;
+
+    if (memcaps == NULL) {
+        memcaps = new_memcap;
+    } else {
+        new_memcap->next = memcaps;
+        memcaps = new_memcap;
+    }
+
+    return 1;
+}
+
+void MemcapListFreeList(void)
+{
+    MemcapList *instance = memcaps;
+
+    while (instance) {
+        MemcapList *next = instance->next;
+        SCFree(instance);
+        instance = next;
+    }
+}
+
+MemcapList *MemcapListGetElement(int index)
+{
+    MemcapList *node = memcaps;
+    int i;
+
+    for (i = 0; node != NULL; i++) {
+        if (i == index) {
+            return node;
+        }
+        node = node->next;
+    }
+    return NULL;
+}
+
+void DisplayMemcaps(void)
+{
+    int i;
+    MemcapList *memcap = NULL;
+
+    for (i = 0; (memcap = MemcapListGetElement(i)); i++) {
+        if (memcap->option) {
+            const char *memcapval;
+            if (ConfGetValue(memcap->option, &memcapval) == 1) {
+                printf("%s = %s\n", memcap->option, memcapval);
+            }
+        }
+    }
+}
+
+void GlobalMemcapInitConfig(void)
+{
+    const char *conf_val = NULL;
+    SC_ATOMIC_INIT(global_memcap);
+
+    MemcapListRegisterMemcap("global", "global-memcap",
+                             GlobalMemcapSetValue, GlobalMemcapGetValue,
+                             NULL);
+
+    if ((ConfGet("global-memcap", &conf_val)) == 1)
+    {
+        if (conf_val == NULL) {
+            SCLogError(SC_ERR_INVALID_YAML_CONF_ENTRY, "Invalid value for "
+                       "global-memcap: NULL");
+            exit(EXIT_FAILURE);
+        }
+        uint64_t global_memcap_copy;
+        if (ParseSizeStringU64(conf_val, &global_memcap_copy) < 0) {
+            SCLogError(SC_ERR_SIZE_PARSE, "Error parsing global-memcap "
+                       "from conf file - %s. Killing engine",
+                       conf_val);
+            exit(EXIT_FAILURE);
+        } else {
+            SC_ATOMIC_SET(global_memcap, global_memcap_copy);
+            g_use_global_memcap = TRUE;
+            SCLogConfig("global memcap is set: %"PRIu64"", SC_ATOMIC_GET(global_memcap));
+        }
+    }
+}
+
+bool GlobalMemcapEnabled(void)
+{
+    return g_use_global_memcap;
+}
+
+uint64_t GlobalMemcapGetValue(void)
+{
+    uint64_t memcapcopy = SC_ATOMIC_GET(global_memcap);
+    return memcapcopy;
+}
+
+int GlobalMemcapSetValue(uint64_t size)
+{
+    SC_ATOMIC_SET(global_memcap, size);
+    return 1;
+}
+
+void GlobalMemcapReached(uint64_t value, const char *name, bool allow_unlimited)
+{
+    if (g_use_global_memcap == FALSE) {
+        return;
+    }
+
+    if (g_memcaps_sum + value > SC_ATOMIC_GET(global_memcap)) {
+        SCLogError(SC_ERR_INVALID_VALUE, "Global memcap value needs to be "
+                   "increased to fit \'%s\' value", name);
```

Comment:
Suggest using `FatalError` instead.

Suggest giving info on current value, current requested value (to be added) in addition to the name.